### PR TITLE
Change Span specification to focus more on capabilities.

### PR DESCRIPTION
**Pull request by bogdandrutu**
This PR changes a subset of the Span API to the agreed model in #165. The rest of the changes will be added in a different PR, trying to keep this PR small so feedback can be addressed faster.

Updates #165

### CODE REVIEW

File: specification/tracing-api.md [hunk 1]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+- An API to record links where the `Link` properties are passed as arguments.
+This SHOULD be called `AddLink`.
+- An API to record lazy initialized links. This can be implemented by providing
+an `Link` interface or a concrete `Link` definition and a `LinkFormatter`. If
+the language supports overloads then this SHOULD be called `AddLink` otherwise
+`AddLazyLink` may be consider.
 
-API MUST also provide an overload that accepts a [`Link` interface](#link). This
-overload allows instrumentation to supply a lazily calculated `Link`.
+#### Set Status
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: I did not get to the Span API. I will clarify in the followup PR.

File: specification/tracing-api.md [hunk 2]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+- (Required) The attribute value, which must be either a string, a boolean
+value, or a numeric type.
 
-- The attribute key, which must be a string.
-- The attribute value, which must be either a string, a boolean value, or a
-  numeric type.
+The Span interface MUST provide:
+- An API to set attributes where the attribute properties are passed as
+arguments. This SHOULD be called `SetAttribute`. To avoid extra allocations some
+implementations may decided to offer on API for every possible value types.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: if supports overloads then same method/function name can be used, otherwise something like setStringAttribute, setBoolAttribute, setLongAttribute.

File: specification/tracing-api.md [hunk 3]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+the values must be either a string, a boolean value, or a numeric type.
 
-Optional parameters:
+The `Event` should be an immutable type.
 
-- One or more key:value pairs, where the keys must be strings and the values may
-  be string, booleans or numeric type.
+The Span interface MUST provide:
+- An API to record events where the `Event` properties are passed as arguments.
+This SHOULD be called `AddEvent`.
```
Comment (reyang) [inline specification/tracing-api.md]: It is a bit unclear to me whether `AddEvent` should support just one event, or multiple events.
It would be great if we can have a simplified use of singular vs. plural form. For example `An API to record a single event ...`. (and it is okay to have `Add Events` as the title as users can call `AddEvent` API several times (from "capability" perspective).

File: specification/tracing-api.md [hunk 4]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
 
-- Map of attributes associated with this link. Attributes are key:value pairs
-  where hey is a string and value is one of string, boolean and numeric.
+The Span interface MUST provide:
+- An API to record links where the `Link` properties are passed as arguments.
+This SHOULD be called `AddLink`.
+- An API to record lazy initialized links. This can be implemented by providing
+a `Link` interface or a concrete `Link` definition and a `LinkFormatter`. If
+the language supports overloads then this SHOULD be called `AddLink` otherwise
+`AddLazyLink` may be consider.
```
Comment (yurishkuro) [inline specification/tracing-api.md]: is there a ticket describing the use case for "lazy links"?

File: specification/tracing-api.md [hunk 5]
```diff
@@ -23,19 +23,15 @@ Table of Content
...
     * [IsRecordingEvents](#isrecordingevents)
-    * [SetAttribute](#setattribute)
-    * [AddEvent](#addevent)
-    * [AddLink](#addlink)
-    * [SetStatus](#setstatus)
+    * [Set Attributes](#set-attributes)
+    * [Add Events](#add-events)
+    * [Add Links](#add-links)
+    * [Set Status](#set-status)
     * [UpdateName](#updatename)
```
Comment (songy23) [inline specification/tracing-api.md]: This one will be updated in a separate PR?

File: specification/tracing-api.md [hunk 6]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
 attributes"](../semantic-conventions.md) that have prescribed semantic meanings.
 
-#### AddEvent
-
-Add an `Event` to `Span`.
+#### Add Events
 
-Required parameters:
+A `Span` MUST have the ability to add events. Events have a time associated
+with the moment when they are added to the `Span.
```
Comment (c24t) [inline specification/tracing-api.md]:
```suggestion
with the moment when they are added to the `Span`.
```

File: specification/tracing-api.md [hunk 7]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
 
-- Name of the event.
+An `Event` is defined by the following properties:
+- (Required) Name of the event.
+- (Optional) One or more key:value pairs, where the keys must be strings and
+the values must be either a string, a boolean value, or a numeric type.
 
-Optional parameters:
+The `Event` should be an immutable type.
```
Comment (c24t) [inline specification/tracing-api.md]: Nit: which "should"s get the RFC 2119 treatment?

File: specification/tracing-api.md [hunk 8]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
 
-- One or more key:value pairs, where the keys must be strings and the values may
-  be string, booleans or numeric type.
+The Span interface MUST provide:
+- An API to record events where the `Event` properties are passed as arguments.
+This SHOULD be called `AddEvent`.
+- An API to record lazy initialized events. This can be implemented by providing
```
Comment (c24t) [inline specification/tracing-api.md]:
```suggestion
- An API to record lazily initialized events. This can be implemented by providing
```

Or "lazy-initialized".

File: specification/tracing-api.md [hunk 9]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
-same or different trace. See [Links description](../terminology.md#links-between-spans).
+#### Add Links
 
-Required parameters
+A `Span` MUST have the ability to add links to other `Span`s. Linked `Span`s
+can be from the same or a different trace. See [Links
+description](../terminology.md#links-between-spans).
 
-- `SpanContext` of the `Span` to link to
+An `Link` is defined by the following properties:
```
Comment (c24t) [inline specification/tracing-api.md]:
```suggestion
A `Link` is defined by the following properties:
```

File: specification/tracing-api.md [hunk 10]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+- (Required) `SpanContext` of the `Span` to link to.
+- (Optional) One or more key:value pairs, where the keys must be strings and
```
Comment (c24t) [inline specification/tracing-api.md]: This could be `Attributes` too.

**Comment by tsloughter** [comment]
This reads much better :). 

I was writing up a whole thing about how in Erlang, with immutable variables, SpanContext maps fine to how the API is defined but the Span's themselves don't. The actual spans are stored in a mutable table keyed on the spancontext. so calls like `set_status` take the span context and update the status in the table without ever reading the whole span out.. But this PR seems to resolve all that and fits much better.

File: specification/tracing-api.md [hunk 11]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+#### Add Links
 
-Required parameters
+A `Span` MUST have the ability to add links to other `Span`s. Linked `Span`s
+can be from the same or a different trace. See [Links
+description](../terminology.md#links-between-spans).
 
-- `SpanContext` of the `Span` to link to
+An `Link` is defined by the following properties:
+- (Required) `SpanContext` of the `Span` to link to.
```
Comment (reyang) [inline specification/tracing-api.md]: Is this the whole `SpanContext` (including `TraceOptions` and `Tracestate`), or just the `TraceId` and `SpanId`?

File: specification/tracing-api.md (hunk 5 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: yes.

File: specification/tracing-api.md (hunk 11 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: Everything.

File: specification/tracing-api.md [hunk 12]
```diff
@@ -282,14 +278,15 @@ Returns the newly created `Span`.
 With the exception of the method to retrieve the `Span`'s `SpanContext` and
 recording status, none of the below may be called after the `Span` is finished.
 
-#### GetContext
+#### Get Context
 
-Retrieve the `Span`s `SpanContext`
+A `Span` MUST have the ability to return the `SpanContext` associated with it.
+The returned value MUST be the same for the entire Span lifetime.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: I think the current PR does not change the behavior. I do understand your requirement but we should have a separate discussion about this. Filed #189 to track this separately.

File: specification/tracing-api.md [hunk 13]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
+#### Add Events
 
-Required parameters:
+A `Span` MUST have the ability to add events. Events have a time associated
+with the moment when they are added to the `Span.
 
-- Name of the event.
+An `Event` is defined by the following properties:
+- (Required) Name of the event.
+- (Optional) One or more key:value pairs, where the keys must be strings and
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: done.

File: specification/tracing-api.md [hunk 14]
```diff
@@ -300,63 +297,78 @@ There should be no parameter.
...
 keys"](../semantic-conventions.md) which have prescribed semantic meanings.
 
-#### AddLink
-
-Adds a link to another `Span` from this `Span`. Linked `Span` can be from the
-same or different trace. See [Links description](../terminology.md#links-between-spans).
+#### Add Links
 
-Required parameters
+A `Span` MUST have the ability to add links to other `Span`s. Linked `Span`s
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: done.

File: specification/tracing-api.md (hunk 4 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: This is mostly for the attributes, for example the values may be lazily formatted.

File: specification/tracing-api.md (hunk 6 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: done

File: specification/tracing-api.md (hunk 10 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: done.

File: specification/tracing-api.md [hunk 15]
```diff
@@ -282,14 +278,15 @@ Returns the newly created `Span`.
...
+The returned value MUST be the same for the entire Span lifetime.
 
-There should be no parameter.
-
-Returns the `SpanContext` for the given `Span`. The returned value may be
-used even after the `Span` is finished.
+The Span interface MUST provide:
+- An API that returns the `SpanContext` for the given `Span`. The returned value
+may be used even after the `Span` is finished. This SHOULD be called
+`GetContext`.
```
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply specification/tracing-api.md]: done.

File: specification/tracing-api.md (hunk 3 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: Done.

File: specification/tracing-api.md (hunk 7 above)
Comment (bogdandrutu) [reply specification/tracing-api.md]: Done.

**Comment by bogdandrutu** [comment]
Waiting for other approves before I will merge.

**Comment by bogdandrutu** [comment]
I will wait until tomorrow to merge this, if no comments :)
